### DISCUSSION TITLE
Replace Base Math Functionality With GLM

### DISCUSSION THREAD

**Discussion by kklouzal** [recastnavigation/recastnavigation#541] [Ideas]
I propose replacing the `Base Math` implementation with [`OpenGL Mathematics (GLM)`](https://github.com/g-truc/glm).

GLM is a very well established, feature rich, highly performant, header only, mathematics library.

A direct quote from their website:
> This library works perfectly with OpenGL but it also ensures interoperability with other third party libraries and SDK. It is a good candidate for software rendering (raytracing / rasterisation), image processing, physic simulations and any development context that requires a _simple and convenient mathematics library_.

Why?
- Reduce Recast/Detour overall codebase making it easier to manage, maintain, and improve upon in the future.
- Couple Recast/Detour more tightly with applications that are using it. Functions that accept arguments and return values will be directly interchangeable with applications that also use GLM for their mathematics.
- Performance. GLM boasts performance overall providing native internal SIMD support on mathematical operations.
- GLM is stable, future versions and updates promise to not break current implementations.
- C98 compatibility assures portability between any application on any platform.
- GLM has a _huge_ community of contributors dedicated to improving and maintaining the library which offers Recast/Detour a top-of-the-line mathematics library that is highly stable, feature rich, and bug free.
- [SoA style access to basic data types](https://glm.g-truc.net/0.9.1/api/a00240.html#_details) should mean no major design changes would be required to swap to GLM.

While I understand this would require a substantial amount of upfront work, the end result will provide many enhancements to the speed, usability, maintainability, and future improvement of Recast/Detour.

**Comment by grahamboree** [comment]
Thanks for the proposal!  There's a lot of upsides to using something like this, and I think you hit on the major ones above.  As with anything though, there's tradeoffs:

- GLM relies heavily on templates and template metaprogramming which can slow compile times.  It also uses variable shadowing which makes inspecting values in debuggers less ergonomic.
- More complicated build process. Would require embedding a version of GLM in the Recast repo, or embedding as a git submodule, which is finicky.
- Adds a second license (MIT) alongside Recast's ZLIB, though they're similar enough it's probably not a huge deal.
- Recast tries to position itself as codebase-agnostic an un-opinionated as possible.  This is important to ensure it's as easy and straightforward as possible to integrate into any game or game engine.  Using 3rd party libraries starts to form opinions about how people should be writing code to interact with Recast.  This may necessitate glue code between the client game and Recast to convert between game types and recast types.  More rationale here: https://github.com/nothings/stb/blob/master/docs/stb_howto.txt

Overall I think you're on to something here though.  There's a lot of places in recast where having a simple POD abstraction like 
```
struct rcVec3 {
float x, y, z;
};
```
would make the code easier to read and less prone to error.  This kind of POD abstraction would allow us to cast vertex float arrays into `rcVec3` arrays for ease of iteration and more clear value access:

```
float[] vertices = ...;
int numVertices = ...;
for (int vertexIndex = 0; vertexIndex < numVertices; ++vertexIndex)
{
    rcVec3& vertex = ((rcVec3*)(vertices))[i];
    vertex.x += 1;
    vertex.y += 2;
    vertex.z += 3;
    
    // rather than...
    
    // vertices[i * 3 + 0] += 1;
    // vertices[i * 3 + 1] += 2;
    // vertices[i * 3 + 2] += 3;
}
```

Changing the code to use an abstraction like this poses it's own stability risks since it's pretty cross-cutting, but in the long run the project's probably better for it.  I'd be keen to see a proposal for something along these lines.  The key constraints here though are that it needs to have zero or near-zero runtime perf overhead, and it should be limited to the minimal level of implementation to address the readability and maintainability issues present.
